test/dynamo/test_higher_order_ops.py

```diff
@@ -282,13 +282,203 @@ def false_fn2(x, y):
+
+            def forward(self, x):
+                def true_fn(x):
+                    self.buffer += 1
+                    return self.buffer.sum() + x.sum()
+
+                def false_fn(x):
+                    return (x - 1).sum()
+
+                return control_flow.cond(x.shape[0] > 4, true_fn, false_fn, [x])
+
+        mod_for_compile = torch.compile(Foo(), backend=cnt, dynamic=True)
+        mod_for_eager = Foo()
+
+        actual = mod_for_compile(torch.ones(6, 4))
+        ref = mod_for_eager(torch.ones(6, 4))
+        self.assertEqual(actual, ref)
+
+        actual = mod_for_compile(torch.ones(3, 4))
+        ref = mod_for_eager(torch.ones(3, 4))
+        self.assertEqual(actual, ref)
+
+        self.assertExpectedInline(
+            backend.graphs[0].code.strip(),
+            """\
+def forward(self, s0 : torch.SymInt, L_x_ : torch.Tensor):
+    l_x_ = L_x_
+    size = l_x_.size();  l_x_ = None
+    getitem = size[0];  size = None
+    gt = getitem > 4;  getitem = None
+    return (gt,)""",
+        )
+
+    def test_cond_free_variable_in_both_branches(self):
+        backend = EagerAndRecordGraphs()
+        cnt = CompileCounterWithBackend(backend)
+
+        z = torch.ones(4, 4)
+
+        class Foo(torch.nn.Module):
+            def __init__(self):
+                super().__init__()
+                self.register_buffer("buffer", torch.ones(6, 4))
+
+            def forward(self, x, y):
+                def true_fn(x):
+                    return x.sum() + self.buffer.sum() + z.sum()
+
+                def false_fn(x):
+                    return x.sum() - z.sum() - self.buffer.sum()
+
+                return control_flow.cond(y, true_fn, false_fn, [x])
+
+        mod_for_compile = torch.compile(
+            Foo(), backend=cnt, dynamic=True, fullgraph=True
+        )
+        mod_for_eager = Foo()
+
+        self.assertEqual(
+            mod_for_compile(torch.tensor(True), torch.tensor(5)),
+            mod_for_eager(torch.tensor(True), torch.tensor(5)),
+        )
+
+        for node in backend.graphs[0].graph.nodes:
+            if node.op == "call_function" and node.target == control_flow.cond:
+                _, _, _, operands = node.args
+                # Each branch takes 5 inputs (x, true_buffer, true_z, false_buffer, false_z)
+                self.assertEqual(len(operands), 5)
+            if node.op == "get_attr":
+                if str(node.target) in ("cond_true_0, cond_false_0"):
+                    num_placeholders = len(
+                        [
+                            node
+                            for node in getattr(
+                                backend.graphs[0], str(node.target)
+                            ).graph.nodes
+                            if node.op == "placeholder"
+                        ]
+                    )
+                    self.assertEqual(num_placeholders, 5)
+
+    def test_cond_side_effect_in_one_branches(self):
```

Comment:
Maybe mark this with an `expectedFailure` and instead `self.assertEqual(len(backend.graphs), 0)`? Or add a comment about this not supposed to be supported but is at the moment.